python/ray/serve/_private/deployment_state.py

```diff
@@ -2813,6 +2825,28 @@ def update(self) -> bool:
 
         return any_recovering
 
+    def _handle_scheduling_request_failures(
+        self,
+        deployment_id: DeploymentID,
+        scheduling_requests: List[ReplicaSchedulingRequest],
+    ):
+        """Updates internal datastructures when replicas fail to be scheduled."""
+        failed_replicas: List[ReplicaID] = []
+        for scheduling_request in scheduling_requests:
+            if (
+                scheduling_request.status
+                == ReplicaSchedulingRequestStatus.PLACEMENT_GROUP_CREATION_FAILED
+            ):
+                failed_replicas.append(scheduling_request.replica_id)
+                self._deployment_states[deployment_id].record_replica_startup_failure(
```

Comment:
Oh now I just realized there might be an edge cace for calling `record_replica_startup_failure()` here. If there are ever partial success, we will stop tracking failures and log those error. Do we think that's expected behavior?

Reply:
Also feel maybe we don't need to call this multiple times since it's the same error and it's getting overrode anyways. Maybe just need to do it once if there are any `failed_replicas`?

Reply:
> If there are ever partial success, we will stop tracking failures and log those error

Could you elaborate? What do you mean by we will stop tracking failures?

> Maybe just need to do it once

The function also increments a counter that tracks the number of times a replica failed to start, which is why we do it multiple times.

Reply:
Basically when the case when `self._replica_constructor_retry_counter` is less than 0. We will no longer set the retry counter and the error message

Reply:
Ah gotcha, that's expected behavior. It matches the current behavior. Essentially, if one of the replicas starts up successfully, then we assume that the other replicas will start up eventually and leave it to the controller to ensure it happens.

Reply:
Cool in that case LGTM :)